lsコマンドのプログラムを作成

junohm410 opened the pull request:
課題のlsコマンドのプログラムを作成し、ディレクトリに追加しました。

junohm410 committed:
lsコマンドのプログラムを作成

課題のlsコマンドのプログラムを作成し、ディレクトリに追加しました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+end
+
+files_of_current_directory = Dir.glob('*')
```
せっかくなので、処理は基本的にメソッド内に書くようにしましょう。
メソッド外の処理は「最初のメソッドの呼び出し」のみにできると良いですね 👍

yuuu replied:
変数名が少し具体的過ぎると感じました。今後の機能拡張で `'*'` 以外を対象とする可能性もあるので、ここはシンプルに `files` で良いのではないでしょうか？

その他の変数も全体的に具体的過ぎるので、少し抽象度を上げてみましょう。もちろん、抽象的過ぎても良くないので、ちょうどいい塩梅を模索してみてください。

junohm410 replied:
メソッド外の処理を「最初のメソッドの呼び出し」のみになるようにしました🙏

変数名ですが、「わからないよりは長い方がいい」と考えて初回提出時のようにしていましたが、いき過ぎていたようです🤧
どうしても具体でつけてしまう傾向があるみたいで、全体的に自信がないですが自分なりに抽象度を上げてみました。「特にこのあたりがよくない」というところがあればご指摘いただけると大変ありがたいです。よろしくお願いいたします🙇‍♂️

yuuu replied:
変数名、抽象的過ぎず具体的過ぎないちょうどいいレベル感になりましたね ✨ 
コードを読みやすくなりました！

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,42 @@
+files_of_current_directory = Dir.glob('*')
+number_of_characters_in_the_longest_filename = files_of_current_directory.map(&:size).max
+
+number_of_displayed_rows = get_number_of_displayed_rows(files_of_current_directory)
```
今回はクラスを使っていないのでRubocopの指摘の対象外となっていますが、 [AccessorMethodName](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Naming/AccessorMethodName) というルールが存在します。簡単に説明すると「余分な `get_` 」はつけないようにしましょう、というルールです。

このメソッド名は `get_` を省略しても `number_of_displayed_rows(files_of_current_directory)` となるので意味は通じるはずです。変数名とメソッド名が被ってしまう点は何かしら対策を検討ください。

junohm410 replied:
こちら承知いたしました。余分な`get_`は使わないように注意したいと思います。
getを使っていたメソッドの名前を修正しました。

> 変数名とメソッド名が被ってしまう点は何かしら対策を検討ください。

最後のメソッドの中で、自分なりに考えて`lines_per_print = number_of_displayed_rows(files)`としてみましたが、やはり同じことを言い換えただけでちょっと自信がありません……🙇‍♂️適切でなければご指摘いただけると幸いです。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,42 @@
+  number_of_files = files.size
+  quatient_and_remainder = number_of_files.divmod(3)
+
+  if quatient_and_remainder[1].zero?
```
ifの場合とelseの場合で `quatient_and_remainder[0]` が重複しているので、重複をなくせないか検討してみてください。

junohm410 replied:
`if - else`ではなく、後置修飾で`unless`を使う形で修正しました。ただ、こちらでも`quotient`を2回使ってはいるので、これでも重複ということになるかなと思い自信がありません。。
式の戻り値を使ったり別の演算子を使うことで`quotient`を1回で済ませる方法を自分なりに考えてみましたが、現時点では思い浮かびませんでした🙇‍♂️
何か方法論のヒントなどがあれば、（可能であれば）教えていただけるとありがたいです。どうぞよろしくお願いいたします🙇‍♂️

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,42 @@
+
+def get_array_of_files_used_for_printing(files, number_of_rows)
+  array_of_files = Array.new(number_of_rows) { [] }
+  files.each_with_index do |file, i|
```
ここの処理は [Array#transpose](https://docs.ruby-lang.org/ja/latest/method/Array/i/transpose.html) を使うともっと楽できますね

junohm410 replied:
ご教示、ありがとうございます！マニュアルを読み使用し修正しました。
正しい行列の形になっていないとエラーを起こすメソッドとのことなので、最終列に空きが出る場合は空文字を代入するという処理を入れています。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,42 @@
+end
+
+def print_files(array_of_files, number_of_characters)
+  array_of_files.each do |array|
```
`array` だと抽象的すぎて、他の配列と名前が衝突してしまうのでここは少し具体的な名前を検討しましょう。

junohm410 replied:
その通りでした🙇‍♂️ご指摘ありがとうございます。
ここに入るのは1行分のファイル群なので、`files`に変更しました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,42 @@
+  array_of_files.each do |array|
+    array.each.with_index(1) do |file, i|
+      if i == array.size
+        print file
```
確かに右端のファイルは `ljust` する必要はないのですが、 `ljust` しても副作用はほとんどないですし、分岐が増えてしまうデメリットのほうが大きいですね。すべて `print "#{file.ljust(number_of_characters)} "` としてはどうでしょうか？

junohm410 replied:
ご指摘、ありがとうございます🙇‍♂️
もし右端以外のファイルで極端に長いファイル名があったときに、ターミナル上で右端にも長い空白ができてしまうのがどうかなと思い分岐を分けていました。
ただ、そうだとしても今後も機能を改修するのにあたり分岐が増えるのはデメリットかなと感じましたので、分岐させない形に修正いたしました。

junohm410 committed:
処理をメソッド内に集約し、メソッドの呼び出しは最後の1行だけにする

junohm410 committed:
ファイルの二次元配列をprint用に整形するのにtransposeメソッドを使うように修正

junohm410 committed:
最後にファイルをprintする際に、右端をljustするかどうかの条件分岐を無くすように変更

junohm410 committed:
メソッド名、変数名を全体的に抽象化するように修正

junohm410 committed:
表示行数を求めるメソッド内で後置修飾を使うように修正

junohm410 commented:
レビューありがとうございました🙇‍♂️
ご指摘いただいた内容をもとに修正いたしましたので、改めてご確認いただけますと幸いです。よろしくお願いいたします。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,41 @@
+  files = Dir.glob('*')
+  width_per_file = files.map(&:size).max
+
+  lines_per_print = number_of_displayed_rows(files)
```
変数名はちょっとだけ抽象化して `number_of_rows` のような名前でも良いかもしれません 👍 
あと、この程度の長さであれば、次の行とまとめて

```ruby
files = format_files(files, number_of_displayed_rows(files))
```

とするのも手です。

junohm410 replied:
ご確認、ありがとうございます。
メソッドの引数に直接メソッド呼び出しを書くとシンプルにできるときがある、ということを頭に持っておきたいと思います🙏
今回は`files = format_files(files, number_of_displayed_rows(files))`で修正いたしました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,41 @@
+  number_of_files = files.size
+  quotient, remainder = number_of_files.divmod(3)
+
+  quotient += 1 unless remainder.zero?
```
この書き方でも良いです 👍 

以下は参考まで。

- [Float#ceil](https://docs.ruby-lang.org/ja/latest/method/Float/i/ceil.html) を使うともっとスッキリ書けると思います 😄 
- さらに、Ruby 3.2以降では [Integer#ceildiv](https://docs.ruby-lang.org/ja/latest/method/Integer/i/ceildiv.html) という、今回にうってつけなメソッドもあります。

junohm410 replied:
「あまりが出たら1を足して……」という最初のアイデアにひきずられていて、「小数点以下/端数をくりあげた整数が欲しい」ということに気づけませんでした🤧
こんなメソッドがあるんですね。
新しいことをしたいときは、「今ほしい処理が既存のメソッドにないだろうか？」と疑いリファレンスマニュアルをのぞく努力をもっとしていこうと思います🙇‍♂️

今回は`ceildiv`を使用して修正しました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,41 @@
+
+def number_of_displayed_rows(files)
+  number_of_files = files.size
+  quotient, remainder = number_of_files.divmod(3)
```
`3` という数字がマジックナンバーになっているので、ここは定数にしておくと良いですね 👍

junohm410 replied:
すみません、その通りでした🙇‍♂️マジックナンバー、注意します。
プログラムの冒頭で表示列数を設定する定数を定義しました。

junohm410 committed:
2回目のレビュー指摘点を修正

表示行数を求めるメソッド内でceildivメソッドを使用する、マジックナンバーの修正、display_files内の記述をスリム化する、3点の修正を行いました。

junohm410 commented:
レビュー、ありがとうございました🙇‍♂️
修正いたしましたので、ご確認いただけますと幸いです。よろしくお願いいたします。
